### PR TITLE
Add GetLParam<T> and GetLParamRef<T>

### PR DESCRIPTION
**Pull request by sharwell**
This pull request implements the new `GetLParam<T>` and `GetLParamRef<T>` (#792) as internal methods, and uses them to avoid marshaling overhead for existing value types.

The first commit is shared with #818.

### CODE REVIEW

**Comment by danmosemsft** [comment]
cc @JeremyKuhne

**Comment by zsd4yr** [comment]
honestly LGTM; I don't really understand the implications however...

File: src/Common/src/UnsafeNativeMethods.cs [hunk 1]
```diff
@@ -1608,6 +1608,15 @@ public static bool DeleteCompatibleDC(HandleRef hDC)
             return regionRects;
         }
 
+        public static ref T PtrToRef<T>(IntPtr ptr)
```
Comment (JeremyKuhne) [inline src/Common/src/UnsafeNativeMethods.cs]: Probably better to use [`Unsafe.AsRef<T>(void* source)`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.asref?view=dotnet-plat-ext-2.1#System_Runtime_CompilerServices_Unsafe_AsRef__1_System_Void__).

**Comment by JeremyKuhne** [comment]
@jkotas, @stephentoub, @jkoritzinsky Any feedback on using this `ref T` approach versus unsafe and `T*` (i.e. `->`) to access struct members in native data? Are there any unobvious pro/cons? Thanks!

**Comment by jkotas** [comment]
I think it is better to stick to regular unmanaged pointers for interop. It makes the intent much more obvious. For example, I do think that the following change is an improvement `NativeMethods.WINDOWPOS* wp = (NativeMethods.WINDOWPOS *)m.LParam;` to `ref NativeMethods.WINDOWPOS wp = ref m.GetLParamRef<NativeMethods.WINDOWPOS>();`.

**Comment by danmosemsft** [comment]
@sharwell do you plan to recast this in terms of pointers, as requested by @jkotas above?

**Comment by sharwell** [comment]
@danmosemsft I'll take a look Monday

**Comment by sharwell** [comment]
> @sharwell do you plan to recast this in terms of pointers, as requested by @jkotas above?

I investigated this change this morning, and my conclusion is I do not believe the change leads to an improved outcome. There is no pointer representation equivalent to `ref readonly`, and I view the `readonly` constraint as a valuable restriction on usage.

File: src/Common/src/UnsafeNativeMethods.cs (hunk 1 above)
Comment (sharwell) [reply src/Common/src/UnsafeNativeMethods.cs]: ➡️ This commit already merged as part of #818. I am submitting a follow-up PR to make this method use `Unsafe.AsRef` for its implementation.

Comment (sharwell) [reply src/Common/src/UnsafeNativeMethods.cs]: Submitted #898

**Comment by jkotas** [comment]
> Conclusion is I do not believe the change leads to an improved outcome.

It is matter of personal preference. Here is a quick summary of pros and cons:

Managed references:
- More overhead (e.g. each instantiation of GetLParamRef is several hundred bytes of memory, conversions from pointers to byrefs turn the GC tracking on that increases size of GCInfo and generates worse code here and there because of that)
- Unsafe is hidden.
- Able to use readonly.

Unmanaged pointers:
- Leaner and faster
- Unsafe is explicit. The unsafe code looks more like C/C++, less like C#.
- Unable to use readonly.

I prefer leaner and faster code with explicit unsafe because of it is much easier to see that is going on. I believe it is also where the CoreCLR/CoreFX repos are leaning towards.

If other folks prefer the opposite and understand consequences of such preference, I do not have a problem with it. Ultimately, it is up to the WinForms maintainers to make the choice.

**Comment by sharwell** [comment]
I believe the perceived magnitude of this downside differs between us:

> ... The unsafe code looks more like C/C++, less like C#.

The pointers don't represent the way I would write (or suggest others write) C#. It doesn't mean it can't be done that way, but if managed references aren't a possibility here it would likely save a bunch of time to go ahead and close this.

> More overhead (e.g. each instantiation of GetLParamRef is several hundred bytes of memory, conversions from pointers to byrefs turn the GC tracking on that increases size of GCInfo and generates worse code here and there because of that)

I would expect this to improve over time as managed references become the preferred style for users.

**Comment by JeremyKuhne** [comment]
> I prefer leaner and faster code with explicit unsafe because of it is much easier to see that is going on.

I agree with this when it comes to direct interop with native code, particularly at these low levels of the stack. I believe there is value in giving uplevel code the best performance possible. On top of performance, the obfuscated "unsafeness" concerns me when it comes to long term maintenance. You can mitigate that somewhat by liberally putting `Unsafe` in methods that do unsafe things. If WinForms is ok with using ref instead I would highly recommend using [`System.Runtime.CompilerServices.Unsafe.AsRef`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.asref?view=dotnet-plat-ext-3.0#System_Runtime_CompilerServices_Unsafe_AsRef__1_System_Void__) directly (or at worst, remove the indirection of `GetLParamRef<T>` or get "Unsafe" somewhere in the name).

> I would expect this to improve over time as managed references become the preferred style for users.

Quite possibly, but we should be very cautious about taking hits based on potential future improvements.

> Ultimately, it is up to the WinForms maintainers to make the choice.

For the record, that isn't me. I'm on the CoreFx team, and I'm here because I've been spending quite a bit of time on interop and I also am very interested in seeing interop improved in WinForms and WPF.

**Comment by sharwell** [comment]
> You can mitigate that somewhat by liberally putting Unsafe in methods that do unsafe things.

I wouldn't be opposed to this. We can't use `Unsafe.AsRef` directly because it doesn't have an overload that takes an `IntPtr`.

**Comment by JeremyKuhne** [comment]
> We can't use Unsafe.AsRef directly because it doesn't have an overload that takes an IntPtr.

Well, not without adding `unsafe` and either casting to `(void*)` or calling `.ToPointer()`. ;P

**Comment by tannergooding** [comment]
> I would expect this to improve over time as managed references become the preferred style for users.

I don't think this will ever get as slim as raw pointers. You also have other issues, such as not being able to accurately handle/represent things like `null`.

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.cs [hunk 2]
```diff
@@ -12842,15 +12842,15 @@ internal void WindowReleaseHandle()
         ///     Handles the WM_NOTIFY message
         /// </devdoc>
         /// <internalonly/>
-        private unsafe void WmNotify(ref Message m) {
-            NativeMethods.NMHDR* nmhdr = (NativeMethods.NMHDR*)m.LParam;
-            if (!ReflectMessage(nmhdr->hwndFrom,ref m)) {
-                if(nmhdr->code == NativeMethods.TTN_SHOW) {
-                    m.Result = UnsafeNativeMethods.SendMessage(new HandleRef(null, nmhdr->hwndFrom), Interop.WindowMessages.WM_REFLECT + m.Msg, m.WParam, m.LParam);
+        private void WmNotify(ref Message m) {
+            ref readonly NativeMethods.NMHDR nmhdr = ref m.GetLParamRef<NativeMethods.NMHDR>();
```
Comment (tannergooding) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.cs]: Is there actually any guarantee that the underlying `nmhdr` is readonly or are you just using this to say the local `WmNotify` shouldn't modify the state?

File: src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs [hunk 3]
```diff
@@ -627,21 +627,19 @@ internal class Indexer {
...
-                        destPtr = new IntPtr(targetData.Scan0.ToInt64() + offsetDest);
-                    }
-                    UnsafeNativeMethods.CopyMemory(new HandleRef(this, destPtr), new HandleRef(this, srcPtr), Math.Abs(targetData.Stride)); 
-                    offsetSrc += sourceData.Stride;
-                    offsetDest += targetData.Stride;
+            for (int i = 0; i < targetData.Height; i++)
+            {
+                IntPtr srcPtr, destPtr;
+                if (IntPtr.Size == 4) {
+                    srcPtr = new IntPtr(sourceData.Scan0.ToInt32() + offsetSrc);
```
Comment (tannergooding) [inline src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs]: `IntPtr` has a `operator +(IntPtr pointer, int offset)` method.

**Comment by zsd4yr** [comment]
hey @sharwell , i kind of fell out of touch here while i was OOF; how is this going?

**Comment by zsd4yr** [comment]
@jkotas , @tannergooding , and @JeremyKuhne  -- could one of you please approve when your perspective is satisfied?

@sharwell pinging again 😄

**Comment by tannergooding** [comment]
@zsd4yr, I think the perspective given so far was that this PR is largely unnecessary and sticking with raw pointers/unsafe code is probably better (e.g: https://github.com/dotnet/winforms/pull/835#issuecomment-487466539).

**Pull request closed without merging**